docs/repo/property-pages/how-to-add-a-new-launch-profile-kind.md

```diff
@@ -103,6 +103,125 @@ And you're done. Projects that import the .targets file will now show this UI wh
+    internal partial class MyPropertyPageProperties : StronglyTypedPropertyAccess
+    {
+        [ExportPropertyXamlRuleDefinition("<Assembly Name>, Version=<Assembly Version>, Culture=<Assembly Culture>, PublicKeyToken=<Assembly Public Key Token>", "XamlRuleToCode:MyPropertyPage.xaml", "Project")]
+        [AppliesTo(MyProject.UniqueCapability)]
+        private object MyRule { get { throw new NotImplementedException(); } }
+
+        [ImportingConstructor]
+        public MyPropertyPageProperties(ConfiguredProject configuredProject)
+            : base(configuredProject)
+        {
+        }
+
+        public MyPropertyPageProperties(ConfiguredProject configuredProject, string file, string itemType, string itemName)
+            : base(configuredProject, file, itemType, itemName)
+        {
+        }
+
+        public MyPropertyPageProperties(ConfiguredProject configuredProject, IProjectPropertiesContext projectPropertiesContext)
+            : base(configuredProject, projectPropertiesContext)
+        {
+        }
+
+        public MyPropertyPageProperties(ConfiguredProject configuredProject, UnconfiguredProject project)
+            : base(configuredProject, project)
+        {
+        }
+    }
+}
+```
+
+### Step 3: Register the assembly codebase
+
+In order for Visual Studio to locate the rule, you must register the codebase of the assembly embedding the rule from the Visual Studio package via the `ProvideCodeBaseAttribute`.
+
+> Note: this is true *even when other MEF components are successfully loaded without a registered codebase*.
+
+```csharp
+using Microsoft.VisualStudio.Shell;
+
+[assembly: ProvideCodeBase(CodeBase = @"$PackageFolder$\Path\To\MyAssembly.dll")]
+
+```
+
+### Step 4: Bind rule properties to launch settings values
+
+Registering the property page allows the settings UI to be visible when the launch profile is selected, but settings will not be read from or written to the launch profile until they are bound to a specific launch profile values via a `ILaunchProfileExtensionValueProvider` implementation.
+
+```csharp
+using Microsoft.Build.Framework.XamlTypes;
+using Microsoft.VisualStudio.ProjectSystem;
+using Microsoft.VisualStudio.ProjectSystem.Debug;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using System.Collections.Immutable;
+
+namespace MyNamespace
+{
+    [ExportLaunchProfileExtensionValueProvider(
+    new[]
+    {
+            ExecutablePathPropertyName
+    },
+    ExportLaunchProfileExtensionValueProviderScope.LaunchProfile)]
+    [AppliesTo(MyProject.UniqueCapability)]
+    internal sealed class MyLaunchProfileExtensionValueProvider : ILaunchProfileExtensionValueProvider
+    {
+        private const string ExecutablePathPropertyName = "ExecutablePath";
+
+        public string OnGetPropertyValue(string propertyName, ILaunchProfile launchProfile, ImmutableDictionary<string, object> globalSettings, Rule rule)
+        {
+            string propertyValue = null;
```

Comment:
It'd be good to use nullable reference types in these code samples. It's easier to remove the annotations than it is to add them, and most folks will be on a new enough compiler.

```suggestion
            string? propertyValue = null;
```

I'd stick `#nullable enable` at the top of these files too, just to make it work regardless of project-level settings.

Reply:
I've ended up removing this section as it really belongs in a different HOW TO.